### PR TITLE
implement node priority and use priority to distribute partitions

### PR DESCRIPTION
**Pull request by jyizheng**
## Summary

Add priority to node definition and use priority to distribute work

## Changelog

- New Feature

## Related Issues

#Issue 163

## Test Plan

curl http://127.0.0.1:8080/v1/cluster/add -X POST -H "Content-Type: application/json" -d '{"name":"PI","address":"192.168.1.106:9091","cpus":4, "priority":255}'

curl http://127.0.0.1:8080/v1/cluster/list -X GET

### CODE REVIEW

**Comment by BohuTANG** [comment]
CI is unhappy. 
By the way, if the PR is not ready for review, it would be better to convert it to draft.
Thanks.

**Comment by jyizheng** [comment]
@BohuTANG  The required tests have passed. Could you review and approve and merge it?

File: query/clusters/src/cluster.rs [hunk 1]
```diff
@@ -39,6 +39,7 @@ impl Cluster {
         let mut node = Node {
             name: n.name.clone(),
             cpus: n.cpus,
+            priority: n.priority,
```
Comment (BohuTANG) [inline query/clusters/src/cluster.rs:42]: Would be better to add some comments about priority, such as what's the rule and how we set the value

Comment (jyizheng) [reply query/clusters/src/cluster.rs:46]: Is the comment good enough?

File: query/src/planners/plan_scheduler.rs [hunk 2]
```diff
@@ -44,18 +45,37 @@ impl PlanScheduler {
...
-        // Align to the cluster numbers.
-        let remainder = partitions.len() % cluster_nums;
-        let chunk_size = (partitions.len() + remainder) / cluster_nums + 1;
-        for chunks in partitions.chunks(chunk_size) {
+        let total_chunks = partitions.len();
+        let mut index = 0;
+        let mut num_chunks_so_far = 0;
+        let mut chunk_size;
+
+        while num_chunks_so_far < total_chunks {
```
Comment (BohuTANG) [inline query/src/planners/plan_scheduler.rs:59]: There are some problems here, although the total chunks is evenly divided according to priority, but it is not assigned according to prigionty.
As an example:
If we have 100 partitions and one cluster with 2-nodes:
node1: {..., priority: 7}
node2: {..., priority: 3}

The partitions distribution should be: 
node1: 7/(7+3)
node2: 3/(7+3)

Comment (BohuTANG) [reply query/src/planners/plan_scheduler.rs:59]: hmm, give a deep check , it seems works as the above comments.
Well,  it's better to add an unit test to show different priority has different partitions.

File: query/clusters/src/cluster.rs (hunk 1 above)
Comment (BohuTANG) [reply query/clusters/src/cluster.rs:46]: LGTM.
255 granularity is too fine? Maybe [0-10] is enough.

Comment (jyizheng) [reply query/clusters/src/cluster.rs:46]: Changed.

File: query/src/planners/plan_scheduler.rs (hunk 2 above)
Comment (jyizheng) [reply query/src/planners/plan_scheduler.rs:59]: I added a unit test called "test_scheduler_plan_with_3_nodes_diff_priority"

File: query/src/tests/service.rs [hunk 3]
```diff
@@ -38,6 +38,30 @@ pub async fn try_create_context_with_nodes(nums: usize) -> FuseQueryResult<FuseQ
...
+    nums: usize,
+    p: &[u8],
+) -> FuseQueryResult<FuseQueryContextRef> {
+    // p is the priority array of the nodes.
+    // Its length of it should be nums.
+    assert_eq!(nums, p.len());
+    let addrs = try_start_service(nums).await?;
+    let ctx = crate::tests::try_create_context()?;
+    for (i, addr) in addrs.iter().enumerate() {
+        println!("i={}, {}", i, p[i] as u8);
```
Comment (BohuTANG) [inline query/src/tests/service.rs:60]: This line I guess is for debug,  remove it if we need it anymore.

Comment (jyizheng) [reply query/src/tests/service.rs]: Yuh, removed

**Pull request closed without merging**